### PR TITLE
New model definition structure

### PR DESCRIPTION
**Pull request by brynpickering**
This is a biggie.

Fixes issue #324
Fixes issue #441 
Fixes issue #337
Fixes issue #362
Fixes issue #263
Fixes issue #230

Adresses discussions #483, #484,

Summary of changes in this pull request:

* Technology definitions flattened as per #484.
* Links moved into technologies (to take advantage of shared `tech_group` inheritance, mainly).
* `parent` strictly one of the abstract base groups.
* Other abstract base groups moved to being referred in `inherit`, with `nodes` having the option to also inherit from `node_groups`.
* `_plus` techs removed as per #483, with the option to include lists of carriers in/out (technically for any technology, but one has to be very careful over which carriers they define their other parameters if doing so) and to invoke `storage` of flows in for all but `demand` technologies (this could be updated to also include demand techs).
* YAML schema to check for the worst offenders. I can't leverage the full power of the schema because of how we build our definitions, with the ability to inherit definitions from `node`/`tech` groups, for instance.
* More complex model data checks (that YAML schema can't handle) moved to a yaml file
* `flow_cap` indexed over carriers now that `conversion_plus` has gone. This required quite a few base math changes and can make it difficult to know what happens when you define a parameter without subsetting its carriers (e.g., if you set `flow_cap_max: 5` for a conversion tech, that will apply to *all* carriers in AND out. You have to use `dims/index` structure to limit it to e.g., one of the output carrier flows).
* Exploding investment cost global expression conditionals into their own global expressions. This reduces build time quite a lot for small models, at the expense of lots of global expressions in the output that might not be useful to the user.

Summary of things still to do:

* Update documentation (sorry sorry sorry)

EDIT: done some

~~* Fix `one_way` parameter so it actually does something (I know how, just need to do it)~~
~~* Add docstrings (sorry sorry)~~
~~* Add test for new helper function which handles referencing a variable/expression that may not be built for a given model (e.g., referencing `storage_cap * cost_storage_cap` when the model includes no techs with storage)~~
~~* Add check for referencing transmission techs in nodes - this should not be allowed.~~
~~* Add check for referencing a different `parent` in different nodes. Although it's technically OK to do this, having to infer the number of dimensions on the resulting array is a pain (ideally they are *only* indexed over `techs`).~~

Things that I won't do in this PR
* Automatically infer `cost_...: 5` as being indexed over `monetary`. I think I'd prefer to avoid this magic.
* Make `lookup` arrays more obvious to the user, rather than a module-level variable in `preprocess/model_data.py`, and add the ability to update it.
* Add the ability to inject extra schema components (e.g., a new parameter with a default etc.)
* Extract parameter data type from the schema and use it to cast the model data arrays. This is a pain right now since most parameters can be strings referring to timeseries data files. This step will be much cleaner if/when we move loading data from file to a different part of the model definition/config (#92).

Reviewer checklist:

- [x] Test(s) added to cover contribution
- [ ] Documentation updated
- [x] Changelog updated
- [x] Coverage maintained or improved

### CODE REVIEW

**Comment by brynpickering** [comment]
The new global expressions (particularly the `flow_in_inc_eff`/`flow_out_inc_eff` ones) increase the memory footprint of models since they keep more unique pyomo objects in memory. For the example models (1yr):

urban scale: 14% increase
national scale: 2% increase

^ I would guess that this is the effect of the extra carriers in the urban scale model.

**Comment by brynpickering** [comment]
The linux py3.9 failure is due to type hinting (I'm surprised it took this long for it to have a problem with my use of pipes for unions). See #519 for proposal to kill py3.9 in favour of py3.12 support.

File: pyproject.toml [hunk 1]
```diff
@@ -36,7 +36,8 @@ output = "reports/coverage/coverage.xml"
 
 [tool.black]
 line-length = 88
-target-version = ['py37', 'py38']
+skip_magic_trailing_comma = true
```
[the remark this answers is not in the document]
Comment (brynpickering) [reply pyproject.toml:39]: If you have this:

```
[very, long, list]
```
black will format it to:


```
[
    very,
    long,
    list,
]
```

If you then remove items but forget to remove the trailing comma, then you'll still get this expanded list. With that param set to true, then:

```
[
    very,
]
```
Will be formatted to:
```
[very]
```

Instead of staying expanded.

File: requirements/base.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
 click >= 8, < 9
+geographiclib >= 2, < 3
```
[the remark this answers is not in the document]
Comment (brynpickering) [reply requirements/base.txt:2]: Because we lifted that code from [here](https://github.com/maurycyp/vincenty) which has issues raised by people that have gone ignored. It also loops to convergance based on a max iteration period which we don't know if it will ever fail on because we can't realistically test all edge cases.

geographiclib has no upstream dependencies, so it seems like a safe (and fast) way to have the same functionality that we outsource maintenance to.

File: src/calliope/core/io.py [hunk 3]
```diff
@@ -20,24 +20,14 @@
...
-from calliope.core.attrdict import AttrDict
+from calliope.attrdict import AttrDict
 
 
 def read_netcdf(path):
     """Read model_data from NetCDF file"""
     with xr.open_dataset(path) as model_data:
         model_data.load()
 
-    calliope_version = model_data.attrs.get("calliope_version", False)
```
[the remark this answers is not in the document]
Comment (brynpickering) [reply src/calliope/core/io.py:32]: It's dealt with centrally in `core.model`

File: src/calliope/preprocess/model_data.py [hunk 4]
```diff
@@ -1,541 +1,726 @@
-# Copyright (C) since 2013 Calliope contributors listed in AUTHORS.
```
[the remark this answers is not in the document]
Comment (brynpickering) [reply src/calliope/preprocess/model_data.py:1]: Oops, quite right.

**Review by sjpfenninger — DISMISSED** [review]

**Review by sjpfenninger — APPROVED** [review]
